From cafe19e009c5837d02afbff90d096f4c802daac0 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Wed, 2 Jun 2021 13:55:39 +0200
Subject: [PATCH] PONRTSYS-9035: Do not check if CMA is reserved on MIPS

When the CMA memory region is reserved over the device tree using
shared-dma-pool which calls rmem_cma_setup() in dma-contiguous.c it uses
the function cma_init_reserved_mem() which checks if the memory region
is reserved. This check fails on MIPS because this is called before the
RAM memory is provided to the Linux memblock system.

It looks like this will be fixed with commit a94e4f24ec83
("MIPS: init: Drop boot_mem_map") upstream, from kernel 5.4, currently
we need this workaround.

Now it is possible to use the reserved-memory functionality to reserve
some CMA memory for a device.
---
 mm/cma.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/mm/cma.c b/mm/cma.c
index 7cb569a188c4..a8d611b5878c 100644
--- a/mm/cma.c
+++ b/mm/cma.c
@@ -178,8 +178,16 @@ int __init cma_init_reserved_mem(phys_addr_t base, phys_addr_t size,
 		return -ENOSPC;
 	}
 
+/*
+ * MIPS uses bootmem to manage early memory allocations and only partly
+ * memblock in kernel 4.9. At this point in the memblock is not fully
+ * initialized on MIPS and this checks fails. I think this should work
+ * better starting with kernel 5.4 and this check should work fine also on MIPS.
+ */
+#ifndef CONFIG_MIPS
 	if (!size || !memblock_is_region_reserved(base, size))
 		return -EINVAL;
+#endif
 
 	/* ensure minimal alignment required by mm core */
 	alignment = PAGE_SIZE <<
